server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java

```diff
@@ -0,0 +1,453 @@
+import static org.elasticsearch.common.util.concurrent.ConcurrentCollections.newConcurrentMap;
+
+public abstract class PeerFinder extends AbstractLifecycleComponent {
+
+    public static final String REQUEST_PEERS_ACTION_NAME = "internal:discovery/zen2/requestpeers";
+
+    // the time between attempts to find all peers
+    public static final Setting<TimeValue> CONSENSUS_FIND_PEERS_DELAY_SETTING =
+        Setting.timeSetting("discovery.zen2.find_peers_delay",
+            TimeValue.timeValueMillis(1000), TimeValue.timeValueMillis(1), Setting.Property.NodeScope);
+
+    private final TimeValue findPeersDelay;
+    private final TimeValue resolveTimeout;
+
+    private final Object mutex = new Object();
+    private final TransportService transportService;
+    private final UnicastHostsProvider hostsProvider;
+    private final FutureExecutor futureExecutor;
+    private final TransportAddressConnector transportAddressConnector;
+    private final Supplier<ExecutorService> executorServiceFactory;
+
+    private final SetOnce<ExecutorService> executorService = new SetOnce<>();
+    private Optional<ActivePeerFinder> peerFinder = Optional.empty();
+
+    PeerFinder(Settings settings, TransportService transportService, UnicastHostsProvider hostsProvider,
+               FutureExecutor futureExecutor, TransportAddressConnector transportAddressConnector,
+               Supplier<ExecutorService> executorServiceFactory) {
+        super(settings);
+        findPeersDelay = CONSENSUS_FIND_PEERS_DELAY_SETTING.get(settings);
+        resolveTimeout = UnicastZenPing.DISCOVERY_ZEN_PING_UNICAST_HOSTS_RESOLVE_TIMEOUT.get(settings);
+        this.transportService = transportService;
+        this.hostsProvider = hostsProvider;
+        this.futureExecutor = futureExecutor;
+        this.transportAddressConnector = transportAddressConnector;
+        this.executorServiceFactory = executorServiceFactory;
+    }
+
+    public Iterable<DiscoveryNode> getFoundPeers() {
+        // getFoundPeersSet() is synchronised so we don't have to
+        return getFoundPeersSet();
+    }
+
+    public boolean foundQuorumFrom(VotingConfiguration votingConfiguration) {
+        // getFoundPeersSet() is synchronised so we don't have to
+        return votingConfiguration.hasQuorum(getFoundPeersSet().stream().map(DiscoveryNode::getId).collect(Collectors.toSet()));
+    }
+
+    Set<DiscoveryNode> getFoundPeersSet() {
+        synchronized (mutex) {
+            return getActivePeerFinder().foundPeers.keySet();
+        }
+    }
+
+    public void activate() {
+        if (lifecycle.started() == false) {
+            logger.debug("ignoring activation, not started");
+            return;
+        }
+
+        // Must not hold lock when calling out to Legislator to avoid a lock cycle
+        assert holdsLock() == false : "Peerfinder mutex is held in error";
+        final DiscoveryNodes lastAcceptedNodes = getLastAcceptedNodes();
```

Comment:
This is tricky because the wake-up has to get the nodes from somewhere. We could pass in a Supplier but it'd still need to do something with locking (maybe this can be avoided with a `LastAcceptedNodesResponder`)? I added a TODO for further discussion.

Reply:
Oh yes, good point. I pushed 2177d0a1a2a.